crates/typst/src/layout/align.rs

```diff
@@ -445,6 +476,150 @@ cast! {
+    type Output = SpecificAlignment<HAlignment, OuterVAlignment>;
+
+    fn add(self, rhs: OuterVAlignment) -> Self::Output {
+        SpecificAlignment::Both(self, rhs)
+    }
+}
+
+impl Add<HAlignment> for OuterVAlignment {
+    type Output = SpecificAlignment<HAlignment, OuterVAlignment>;
+
+    fn add(self, rhs: HAlignment) -> Self::Output {
+        SpecificAlignment::Both(rhs, self)
+    }
+}
+
+impl<H, V> From<SpecificAlignment<H, V>> for Alignment
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn from(value: SpecificAlignment<H, V>) -> Alignment {
+        type FromType<H, V> = SpecificAlignment<H, V>;
+        match value {
+            FromType::H(x) => Alignment::H(x.into()),
+            FromType::V(y) => Alignment::V(y.into()),
+            FromType::Both(x, y) => Alignment::Both(x.into(), y.into()),
+        }
+    }
+}
+
+impl<H, V> Reflect for SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy + Reflect,
+    V: Into<VAlignment> + Copy + Reflect,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn input() -> CastInfo {
+        H::input() + V::input()
+    }
+
+    fn output() -> CastInfo {
+        H::output() + V::output()
+    }
+
+    fn castable(value: &Value) -> bool {
+        H::castable(value) || V::castable(value)
+    }
+}
+
+impl<H, V> IntoValue for SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy + Reflect,
+    V: Into<VAlignment> + Copy + Reflect,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn into_value(self) -> Value {
+        let alignment = match self {
+            Self::H(x) => Alignment::H(x.into()),
+            Self::V(y) => Alignment::V(y.into()),
+            Self::Both(x, y) => Alignment::Both(x.into(), y.into()),
+        };
+        alignment.into_value()
+    }
+}
+
+impl<H, V> FromValue for SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy + Reflect + FromValue,
+    V: Into<VAlignment> + Copy + Reflect + FromValue,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn from_value(value: Value) -> StrResult<Self> {
+        if Alignment::castable(&value) {
+            let align = Alignment::from_value(value)?;
+            let result = match align {
+                Alignment::H(x) => {
+                    let h_value = x.into_value();
+                    Self::H(H::from_value(h_value)?)
+                }
+                Alignment::V(y) => {
+                    let v_value = y.into_value();
+                    Self::V(V::from_value(v_value)?)
+                }
+                Alignment::Both(x, y) => {
+                    let h_value = x.into_value();
+                    let v_value = y.into_value();
+                    Self::Both(H::from_value(h_value)?, V::from_value(v_value)?)
+                }
+            };
+            return Ok(result);
+        }
+        Err(Self::error(&value))
+    }
```

Comment:
I was going to suggest to remove the bounds here, but thinking further, I think you might be able to replace the last three trait implementations by a single `cast!` invocation (take a look at how other types do it).

Reply:
I attempted to use `cast!` but it appears it cannot handle generics properly. For example, the generated code simply writes `impl #foundations::Reflect for` but we want `impl<H, V> Reflect for`.

Reply:
Fair enough, didn't consider that.